examples/graph/wgan-graphs.py

```diff
@@ -0,0 +1,601 @@
+
+                grads = tape.gradient(loss, self.generator.trainable_weights)
+                self.optimizer_generator.apply_gradients(
+                    zip(grads, self.generator.trainable_weights)
+                )
+                self.metric_generator.update_state(loss)
+
+        return {m.name: m.result() for m in self.metrics}
+
+    def _loss_discriminator(self, graph_real, graph_generated):
+        logits_real = self.discriminator(graph_real, training=True)
+        logits_generated = self.discriminator(graph_generated, training=True)
+        loss = tf.reduce_mean(logits_generated) - tf.reduce_mean(logits_real)
+        loss_gp = self._gradient_penalty(graph_real, graph_generated)
+        return loss + loss_gp * self.gp_weight
+
+    def _loss_generator(self, graph_generated):
+        logits_generated = self.discriminator(graph_generated, training=True)
+        return -tf.reduce_mean(logits_generated)
+
+    def _gradient_penalty(self, graph_real, graph_generated):
+        # Unpack graphs
+        adjacency_real, features_real = graph_real
+        adjacency_generated, features_generated = graph_generated
+
+        # Generate interpolated graphs (adjacency_interp and features_interp)
+        alpha = tf.random.uniform([self.batch_size])
+        alpha = tf.reshape(alpha, (self.batch_size, 1, 1, 1))
+        adjacency_interp = (adjacency_real * alpha) + (1 - alpha) * adjacency_generated
+        alpha = tf.reshape(alpha, (self.batch_size, 1, 1))
+        features_interp = (features_real * alpha) + (1 - alpha) * features_generated
+
+        # Compute the logits of interpolated graphs
+        with tf.GradientTape() as tape:
+            tape.watch(adjacency_interp)
+            tape.watch(features_interp)
+            logits = self.discriminator(
+                [adjacency_interp, features_interp], training=True)
+
+        # Compute the gradients with respect to the interpolated graphs
+        grads = tape.gradient(logits, [adjacency_interp, features_interp])
+        # Compute the gradient penalty
+        grads_adjacency_penalty = (1 - tf.norm(grads[0], axis=1)) ** 2
+        grads_features_penalty = (1 - tf.norm(grads[1], axis=2)) ** 2
+        return tf.reduce_mean(
+            tf.reduce_mean(grads_adjacency_penalty, axis=(-2, -1)) +
+            tf.reduce_mean(grads_features_penalty, axis=(-1))
+        )
+
+
+"""
+## Train the model
+
+To save time (if run on a CPU), we'll only train the model for 10 epochs.
+"""
+
+wgan = GraphWGAN(generator, discriminator, discriminator_steps=1)
+
+wgan.compile(
+    optimizer_generator=keras.optimizers.Adam(5e-4),
+    optimizer_discriminator=keras.optimizers.Adam(5e-4),
+)
+
+wgan.fit([adjacency_tensor, feature_tensor], epochs=10, batch_size=16)
+
+"""
+## Sample novel molecules with the generator
+"""
+
+
+def sample(generator, batch_size):
+    z = tf.random.normal((batch_size, LATENT_DIM))
+    graph = generator(z)
```

Comment:
A slight concern here is that `generator(z)` is running eagerly.

Consider using `generator.predict(z)`, or alernatively move the TF part of this function to a `@tf.function`.